### PR TITLE
fix name length bug in restart version

### PR DESCRIPTION
**Pull request by danielhochman**

### CODE REVIEW

**Comment by htuch** [comment]
Yeah, LGTM, thanks, +1 on test. @zuercher Mac snafu?

**Comment by ggreenway** [comment]
Here's the test:
```
diff --git a/test/server/hot_restart_impl_test.cc b/test/server/hot_restart_impl_test.cc
index 6b60a80..2ef43ac 100644
--- a/test/server/hot_restart_impl_test.cc
+++ b/test/server/hot_restart_impl_test.cc
@@ -28,17 +28,32 @@ public:
     }));
     EXPECT_CALL(os_sys_calls_, bind(_, _, _));
 
+    Stats::RawStatData::configureForTestsOnly(options_);
+
     // Test we match the correct stat with empty-slots before, after, or both.
     hot_restart_.reset(new HotRestartImpl(options_, os_sys_calls_));
     hot_restart_->drainParentListeners();
   }
 
+  void TearDown() {
+    // Configure it back so that later tests don't get the wonky values
+    // used here
+    NiceMock<MockOptions> default_options;
+    Stats::RawStatData::configureForTestsOnly(default_options);
+  }
+
   Api::MockOsSysCalls os_sys_calls_;
   NiceMock<MockOptions> options_;
   std::vector<uint8_t> buffer_;
   std::unique_ptr<HotRestartImpl> hot_restart_;
 };
 
+TEST_F(HotRestartImplTest, versionString) {
+  setup();
+  EXPECT_EQ(hot_restart_->version(), Envoy::Server::SharedMemory::version(
+                                         options_.maxStats(), options_.maxStatNameLength()));
+}
+
 TEST_F(HotRestartImplTest, crossAlloc) {
   setup();
 
@@ -87,16 +102,8 @@ public:
   HotRestartImplAlignmentTest() : name_len_(8 + GetParam()) {
     EXPECT_CALL(options_, maxStats()).WillRepeatedly(Return(num_stats_));
     EXPECT_CALL(options_, maxStatNameLength()).WillRepeatedly(Return(name_len_));
-    Stats::RawStatData::configureForTestsOnly(options_);
-    EXPECT_EQ(name_len_, Stats::RawStatData::maxNameLength());
     setup();
-  }
-
-  void TearDown() {
-    // Configure it back so that later tests don't get the wonky values
-    // used here
-    NiceMock<MockOptions> default_options;
-    Stats::RawStatData::configureForTestsOnly(default_options);
+    EXPECT_EQ(name_len_, Stats::RawStatData::maxNameLength());
   }
 
   static const uint64_t num_stats_ = 8;
```

**Comment by ggreenway** [comment]
Also, I'm surprised the hot_restart integration test didn't fail on this bug.

**Comment by zuercher** [comment]
Looks like something went haywire on Circle's end. I'll keep an eye on it, but if it doesn't pass when you fix the format failure, it should be okay to merge anyway.

**Comment by zuercher** [comment]
Hm, it actually restarted on its own (but the replacement build didn't show up for a long time). Looks like it's hung in //test/integration:http2_upstream_integration_test, which should be unrelated to the changes in this PR. I'll take a look at that test and see if there's an intermittent problem.

**Comment by ggreenway** [comment]
Oh, I see why the integration test succeeded.  It doesn't do what I would expect any wrapper script to do: run envoy with --hot-restart-version, and compare that with the output of the admin endpoint of the running envoy.  A hot-restart between two envoys with this bug will succeed.

**Comment by mattklein123** [comment]
@ggreenway @danielhochman if one if you could fix that integration test to test this that would be good. @ggreenway also I forget about this before but can you also add the new CLI options to https://envoyproxy.github.io/envoy/operations/cli.html. Thank you!

**Comment by ggreenway** [comment]
I'm working on writing the integration test change.  I'll make a separate PR for that and the doc changes, unless you want this PR to wait on the integration test.  Either way is fine with me.

File: source/server/hot_restart_impl.cc [hunk 1]
```diff
@@ -101,7 +101,9 @@ std::string SharedMemory::version(size_t max_num_stats, size_t max_stat_name_len
                      max_stat_name_len);
 }
 
-std::string SharedMemory::version() { return version(num_stats_, Stats::RawStatData::size()); }
+std::string SharedMemory::version() {
+  return version(num_stats_, Stats::RawStatData::maxNameLength());
```
Comment (mattklein123) [inline source/server/hot_restart_impl.cc]: Sorry I think I'm confused. How does this matter? What's the difference if it uses the full size?

Comment (mattklein123) [reply source/server/hot_restart_impl.cc]: Or is it just mismatched from what the other thing is returning?

Comment (mattklein123) [reply source/server/hot_restart_impl.cc]: I see, never mind.

**Pull request merged**